### PR TITLE
DMA peripheral

### PR DESCRIPTION
**Pull request by martonbognar**
- Add constants for the attacker peripheral's addresses
- Integrate collecting DMA traces into sancus_step: can be set using `__ss_set_dma_attacker_delay(int delay)`

### CODE REVIEW

File: src/sancus-step/sancus_step.c [hunk 1]
```diff
@@ -72,6 +75,11 @@ void __ss_start(void)
     timer_irqc(INIT_LATENCY);
 }
 
+void __ss_set_dma_attacker_delay(int delay)
```
Comment (jovanbulck) [inline src/sancus-step/sancus_step.c:78]: ideally this call shouldn't be needed I think: 

* initialization of the delay could happen automatically in `ss_init`; now if one doesn't call this function strange things will happen?
* why would one need a non-zero delay argument here?

File: src/sancus-step/sancus_step.c [hunk 2]
```diff
@@ -22,6 +22,8 @@ int __ss_isr_reti_latency = 0;
 int __ss_sm_exit_latency = 0;
 int __ss_isr_interrupted_sm = 0;
 
+int __ss_dma_attacker_delay = 0xFFFF;
```
Comment (jovanbulck) [inline src/sancus-step/sancus_step.c:25]: why this initial value?

File: include/sancus_support/sancus_step.h [hunk 3]
```diff
@@ -110,7 +119,9 @@ void timerA_isr_entry2(void)                                                   \
         "i"(TACTL_ENABLE),                                           /* %10 */ \
         "i"(TACCTL_ENABLE_CONT),                                     /* %11 */ \
         "m"(TACCTL0),                                                /* %12 */ \
-        "i"(TACTL_ENABLE_CONT)                                       /* %13 */ \
+        "i"(TACTL_ENABLE_CONT),                                      /* %13 */ \
+        "m"(__ss_dma_attacker_delay),                                /* %14 */ \
+        "m"(DMA_COUNTDOWN)                                           /* %15 */ \
```
Comment (jovanbulck) [inline include/sancus_support/sancus_step.h:124]: ideally we should move this inline asm to a dedicated `sancus-step.S` file at some point..

**Review by jovanbulck — CHANGES_REQUESTED** [review]

**Comment by jovanbulck** [comment]
looks good :+1: left some comments -- let's try to make it as transparent and robust as possible, without having to make explicit calls preferably

**Comment by martonbognar** [comment]
I'll respond to both the comments here and over at https://github.com/sancus-pma/sancus-examples/pull/13 since they are related to each other.

I added the option of setting a non-zero delay to enable capturing memory traces of the tail ends of theoretical longer-than-16-cycle instructions, or the code that runs while the interrupt handling occurs. This might not be necessary, in that case, I can just make it default to 0 and have a `__ss_enable_dma_traces(void)` function instead or similar.

The delay value `FFFF` works as a "do not activate" magic value in the peripheral (this should probably be documented somewhere...). In case we do not want the DMA requests to fire while running sancus-step, we can just leave the delay untouched, and no DMA requests will be issued by the peripheral.

Another interesting phenomenon I came across while testing this code is captured by the following screenshot:
Screenshot from 2020-03-27 02-18-15

What you can see here is that if the peripheral was activated on the first run (I think this happens in `__ss_init` (with `DMA_PRIORITY CPU`)), it broke the system... I haven't tracked down the reason for this (could be interesting, maybe even a possible attack vector?), but waiting for the user to explicitly activate DMA measurements seemed like a good option to both avoid this problem, and to avoid unnecessary DMA traffic in case the user doesn't want to track memory requests while running sancus_step.

---

Probably the cleanest approach (with the loss of arbitrary delays set by the user, and the added overhead traffic in case we don't want to use the measurements) would be to set the delay (~call `__ss_set_dma_attacker_delay(0)`) at the end of `__ss_init`. This way, the DMA requests would still not fire during the set-up phase, but they would start immediately for the execution of the actual victim module.

**Pull request merged**